File: python/0279完全平方数.py
from math import sqrt

# 组合完全背包
def numSquares(n: int) -> int:
    items=[i*i for i in range(1,int(sqrt(n))+1)]
    
    volume=n+1
    dp=[0 if i==0 else 10e4+1 for i in range(volume)]
    

    for i in items:
        for j in range(i,n+1):
            dp[j]=min(dp[j],dp[j-i]+1)
    
    
    return dp[-1]


print(numSquares(13))

File: csharp/0279完全平方数.cs
public partial class Solution
{
    public int NumSquares(int n)
    {
        int item_size = (int)Math.Sqrt(n);
        int[] items = new int[item_size];
        int[] dp = new int[n + 1];

        for (int i = 0; i < item_size; i++)
        {
            items[i] = (i + 1) * (i + 1);
        }

        for (int i = 1; i <= n; i++)
        {
            dp[i] = int.MaxValue;
        }

        foreach(int i in items)
        {
            for (int j=i;j<=n ;j++)
            {
                dp[j] = Math.Min(dp[j], dp[j - i] + 1);
            }
        }

        return dp[n];
    }
}